source/solvers/free_surface.cc

```diff
@@ -215,6 +223,36 @@ FreeSurface<dim>::assemble_system(
               // indicator
               const double u_mag = std::max(velocity.norm(), 1e-12);
 
+              // Gather the shape functions and their gradient
+              for (unsigned int k : fe_values_fs.dof_indices())
+                {
+                  phi_phase[k]      = fe_values_fs.shape_value(k, q);
+                  grad_phi_phase[k] = fe_values_fs.shape_grad(k, q);
+                }
+
+
+              // Shock capturing viscosity term
+              const double order =
+                this->simulation_parameters.fem_parameters.tracer_order;
+
+              const double vdcdd = (0.5 * h) *
+                                   (velocity.norm() * velocity.norm()) *
+                                   pow(phase_gradients[q].norm() * h, order);
+
+              // Tensor<1, dim> s = velocity / (velocity.norm() + 1e-12);
+              Tensor<1, dim> r =
+                phase_gradients[q] / (phase_gradients[q].norm() + 1e-12);
+
+              // const Tensor<2, dim> k_corr      = (r * s) * outer_product(s,
+              // s);
+              const Tensor<2, dim> rr          = outer_product(r, r);
+              const Tensor<2, dim> dcdd_factor = rr; // - k_corr;
```

Comment:
so remove the variable `rr`? `dcdd_factor = outer_product(r, r);`

Reply:
I'd like to keep the capacity for both approaches.